src/ray/object_manager/object_directory.cc

```diff
@@ -42,63 +67,77 @@ ray::Status ObjectDirectory::GetInformation(const ClientID &client_id,
     success_callback(info);
   }
   return ray::Status::OK();
-};
+}
 
-ray::Status ObjectDirectory::GetLocations(const ObjectID &object_id,
-                                          const OnLocationsSuccess &success_callback,
-                                          const OnLocationsFailure &fail_callback) {
-  ray::Status status_code = ray::Status::OK();
-  if (existing_requests_.count(object_id) == 0) {
-    existing_requests_[object_id] = ODCallbacks({success_callback, fail_callback});
-    status_code = ExecuteGetLocations(object_id);
-  } else {
-    // Do nothing. A request is in progress.
+ray::Status ObjectDirectory::SubscribeObjectLocations(const ObjectID &object_id,
+                                                      const OnLocationsFound &callback) {
+  ray::Status status = ray::Status::OK();
+  if (listeners_.find(object_id) != listeners_.end()) {
+    RAY_LOG(ERROR) << "Duplicate calls to SubscribeObjectLocations for " << object_id;
+    return ray::Status::Invalid("Cannot do things.");
   }
-  return status_code;
-};
-
-ray::Status ObjectDirectory::ExecuteGetLocations(const ObjectID &object_id) {
-  JobID job_id = JobID::nil();
-  // Note: Lookup must be synchronous for thread-safe access.
-  // For now, this is only accessed by the main thread.
-  ray::Status status = gcs_client_->object_table().Lookup(
-      job_id, object_id, [this](gcs::AsyncGcsClient *client, const ObjectID &object_id,
-                                const std::vector<ObjectTableDataT> &data) {
-        GetLocationsComplete(object_id, data);
+  listeners_.emplace(object_id, LocationListenerState(callback));
+  GetLocations(
+      object_id,
+      [this, callback](const std::vector<ray::ClientID> &v,
+                       const ray::ObjectID &object_id) {
+        if (listeners_.count(object_id) > 0) {
```

Comment:
Is the GetLocations is a sync call or async call?
If it is a sync call, we don't need the check here. If it is an async call, it means the listeners_ can be changed during the line, we must add lock to it.